runtime/wasm/src/asc_abi/v0_0_5.rs

```diff
@@ -0,0 +1,267 @@
+    pub fn get<T: AscType>(
+        &self,
+        byte_offset: u32,
+        length: u32,
+        api_version: Version,
+    ) -> Result<Vec<T>, DeterministicHostError> {
+        let length = length as usize;
+        let byte_offset = byte_offset as usize;
+
+        self.content[byte_offset..]
+            .chunks(size_of::<T>())
+            .take(length)
+            .map(|asc_obj| T::from_asc_bytes(asc_obj, api_version.clone()))
+            .collect()
+    }
+}
+
+impl AscType for ArrayBuffer {
+    fn to_asc_bytes(&self) -> Result<Vec<u8>, DeterministicHostError> {
+        let mut asc_layout: Vec<u8> = Vec::new();
+
+        asc_layout.extend(self.content.iter());
+
+        // Allocate extra capacity to next power of two, as required by asc.
+        let total_size = self.byte_length as usize + HEADER_SIZE;
+        let total_capacity = total_size.next_power_of_two();
+        let extra_capacity = total_capacity - total_size;
+        asc_layout.extend(std::iter::repeat(0).take(extra_capacity));
+
+        Ok(asc_layout)
+    }
+
+    fn from_asc_bytes(
+        asc_obj: &[u8],
+        _api_version: Version,
+    ) -> Result<Self, DeterministicHostError> {
+        Ok(ArrayBuffer {
+            byte_length: asc_obj.len() as u32,
+            content: asc_obj.to_vec().into(),
+        })
+    }
+
+    fn content_len(&self, _asc_bytes: &[u8]) -> usize {
+        self.byte_length as usize // without extra_capacity
+    }
+}
+
+/// A typed, indexable view of an `ArrayBuffer` of Asc primitives. In Asc it's
+/// an abstract class with subclasses for each primitive, for example
+/// `Uint8Array` is `TypedArray<u8>`.
+/// Also known as `ArrayBufferView`.
+/// See https://www.assemblyscript.org/memory.html#arraybufferview-layout
+#[repr(C)]
+#[derive(AscType)]
+pub struct TypedArray<T> {
+    // #data -> Backing buffer reference
+    pub buffer: AscPtr<ArrayBuffer>,
+    // #dataStart -> Start within the #data
+    data_start: u32,
+    // #dataLength -> Length of the data from #dataStart
+    byte_length: u32,
+    // Not included in memory layout, it's just for typings
+    ty: PhantomData<T>,
+}
+
+impl<T: AscValue> TypedArray<T> {
+    pub(crate) fn new<H: AscHeap + ?Sized>(
+        content: &[T],
+        heap: &mut H,
+    ) -> Result<Self, DeterministicHostError> {
+        let buffer = class::ArrayBuffer::new(content, heap.api_version())?;
+        let byte_length = content.len() as u32;
+        let ptr = AscPtr::alloc_obj(buffer, heap)?;
+        Ok(TypedArray {
+            buffer: AscPtr::new(ptr.wasm_ptr()), // new AscPtr necessary to convert type parameter
+            data_start: ptr.wasm_ptr(),
+            byte_length,
+            ty: PhantomData,
+        })
+    }
+
+    pub(crate) fn to_vec<H: AscHeap + ?Sized>(
+        &self,
+        heap: &H,
+    ) -> Result<Vec<T>, DeterministicHostError> {
+        self.buffer.read_ptr(heap)?.get(
+            self.data_start - self.buffer.wasm_ptr(), // needed to get offset
```

Comment:
Better comment and use checked subtraction (same for `Array<T>`)

Reply:
Solved!